Upgrade to Dokka 2.1.0

cppwfs opened the pull request:
- Update `org.jetbrains.dokka` plugin from 2.0.0 to 2.1.0
- Remove Jackson 2.15.3 version resolution strategy workaround that was needed to avoid conflicts in Dokka 2.0.0
- Migrate from deprecated V1 tasks (`dokkaHtmlPartial`, `dokkaHtmlMultiModule`) to V2 API (`dokkaGeneratePublicationHtml`, `dokkaGenerate`)
- Replace `dokkaSourceSets.main` configuration block with `dokkaSourceSets.configureEach`
- Update `docsZip` task to depend on new `dokkaGeneratePublicationHtml` task and add explicit dependency on `:spring-integration-core:dokkaGenerate`

<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc).
-->

cppwfs committed:
Upgrade to Dokka 2.1.0

- Update `org.jetbrains.dokka` plugin from 2.0.0 to 2.1.0
- Remove Jackson 2.15.3 version resolution strategy workaround that
  was needed to avoid conflicts in Dokka 2.0.0
- Migrate from deprecated V1 tasks (`dokkaHtmlPartial`,
  `dokkaHtmlMultiModule`) to V2 API (`dokkaGeneratePublicationHtml`,
  `dokkaGenerate`)
- Replace `dokkaSourceSets.main` configuration block with
  `dokkaSourceSets.configureEach`
- Update `docsZip` task to depend on new
  `dokkaGeneratePublicationHtml` task and add explicit dependency on
  `:spring-integration-core:dokkaGenerate`

artembilan reviewed (changes requested):
Thank you for a swift turn around!

Not too much, just couple nit-picks and some concerns.
Can you share, please, with me a screenshot how the final doc looks?

After this we would be able to migrate to Gradle `9.1` finally 😄

cppwfs commented:
<img width="944" height="604" alt="image" src="https://github.com/user-attachments/assets/be682c19-9ac0-4c6d-94e7-2d55aa78491b" />
<img width="1184" height="775" alt="image" src="https://github.com/user-attachments/assets/3e10f9c6-5507-4fdb-8c32-0158dd3adce8" />

cppwfs committed:
Restore external documentation links for Dokka

- Restore external documentation links in spring-integration-core's
  `dokkaSourceSets` using the new `externalDocumentationLinks.
  register()` API instead of deprecated `externalDocumentationLink`
- Fix indentation of spring-integration-core's `dokka` configuration
  block to be properly scoped within the project
- Remove redundant `noJdkLink.set(false)` setting (false is default)
- Remove unnecessary root-level `dokka` configuration that was
  setting module name and output directory without effect
- Remove unused `dokkaGeneratePublicationHtml` task dependency from
  `docsZip` as `dokkaGenerate` already handles the generation
- Simplify dependency chain by relying solely on
  `:spring-integration-core:dokkaGenerate`
- Use root project var to determine root
- Move the out of core module
  This makes the dokka build available to all modules if necessary
- Only run the module in the core module

cppwfs committed:
Optimize Dokka to process selected modules only

- Refactor Dokka configuration to apply only to modules with Kotlin
  sources, reducing build overhead and improving performance for
  modules without Kotlin code.
- Remove deprecated commands

artembilan commented on build.gradle:
```diff
@@ -197,6 +188,8 @@ configure(javaProjects) { subproject ->
     apply plugin: 'kotlin'
     apply plugin: 'kotlin-spring'
     apply plugin: 'io.spring.nullability'
+    apply plugin: 'org.jetbrains.dokka'
```
This is not correct: it is waste of resource to apply the plugin in the sub-project where is nothing with Kotlin.

artembilan reviewed (changes requested):
This is what Spring Framework does:

https://github.com/spring-projects/spring-framework/blob/main/framework-api/framework-api.gradle

And this line is strange for me: https://github.com/spring-projects/spring-framework/blob/main/buildSrc/src/main/java/org/springframework/build/KotlinConventions.java#L40

artembilan reviewed (changes requested):
Looks like there is no way to make it transparent: https://kotlinlang.org/docs/dokka-migration.html#update-documentation-aggregation-in-multi-module-projects.
We have to be explicit with that our module:
```
dependencies {
    dokka(project(':spring-integration-core'))
}
```
in the root project 😢

artembilan reviewed (changes requested):
Strange. I thought I sent my review.
Sorry about that.

cppwfs committed:
Simplify Dokka configuration with auto-discovery

Replace static `dokkaProjects` list with dynamic Kotlin source
detection, eliminating manual maintenance and improving automation.

- Remove `dokkaProjects` list from ext block that required manual
  updates when adding Kotlin modules
- Move Dokka plugin and configuration back into main `javaProjects`
  configure block for simpler structure
- Update `docsZip` task to dynamically find projects with Kotlin
  sources using `javaProjects.findAll {it.file('src/main/kotlin').
  exists()}` instead of referencing static list
- Change output directory API from `buildDirectory.file('kdoc')` to
  `buildDirectory.dir('kdoc')` for semantically correct directory
  reference

cppwfs committed:
Use dokka plugin on required modules
